### PR TITLE
Require end-of-input to be reached for range-based parser invocations

### PR DESCRIPTION
**Pull request by dominiklohmann**
Motivation: We use a custom-made parser-combinator framework for high-performance parsers in VAST, and to combine parsers one needs to be able to resume parsing after partial consumption of input. Experience has shown (e.g., #791) that we don't usually want to silently discard input on the resulting, combined parsers.

Before the change implemented by this PR, EOI checks were required to be done manually by the caller, e.g. by calling `(p >> eoi)(rng, attr)` over `p(rng, attr)`. Outside of combining with the `eoi` parser, there was no way to verify whether the EOI was reached successfully for range-based parser invocations. This has turned out to be very error prone.

For `[begin, end)`-iterator based invocations, nothing changes. It is easy for the caller to verify whether the EOI was reached successfully by simply comparing `begin` and `end`.

Among other things, this PR completely disables the MRT integration, which we knew to be broken, but which silently succeeded in parsing invalid data. It will be revisited at a later point in time.

### CODE REVIEW

**Comment by dominiklohmann** [comment]
[ch13963]

**Comment by dominiklohmann** [comment]
End result:
- CSV map parser is broken; test has been disabled.
- MRT/bgpdump integrations are broken; have been removed (for now). This was known to be broken, it just never failed in the tests.
- A whole bunch of parsers were fixed by the second commit, which strips the NUL byte from parser invocations using a known-length char array. This mostly affected our test suite.

File: libvast/vast/concept/parseable/core/parser.hpp [hunk 1]
```diff
@@ -54,6 +54,15 @@ struct parser {
     return guard_parser<Derived, Guard>{derived(), fun};
   }
 
+  template <size_t N, class Attribute = unused_type>
+  bool operator()(const char (&r)[N], Attribute& a = unused) const {
```
[the remark this answers is not in the document]
Comment (dominiklohmann) [reply libvast/vast/concept/parseable/core/parser.hpp]: I've thought about that and don't think it's a good idea, as it'd break passing an array of bytes. By restricting the type as closely as we can to string literals, we achieve what we want without being too disruptive.

> Looks good, I'm surprised there are not more breaking tests.

Removing this additional operator overload makes a quarter of the tests in the code base fail... That's what I was doing for the first hour before I noticed they were all off-by-one errors on string inputs.

[the remark this answers is not in the document]
Comment (tobim) [reply libvast/vast/concept/parseable/core/parser.hpp]: Ok, so the condition would get a bit more complex. I have no strong opinion either way, so this is fine.

**Pull request merged**